pkg/dedup/merge_iter.go

```diff
@@ -12,30 +12,37 @@ import (
 	"github.com/prometheus/prometheus/tsdb/chunkenc"
 )
 
-const UseMergedSeries = "use_merged_series"
-
 // mergedSeries is a storage.Series that implements a simple merge sort algorithm.
 // when replicas has conflict values at the same timestamp, the first replica will be selected.
 type mergedSeries struct {
 	lset     labels.Labels
 	replicas []storage.Series
+
+	isCounter bool
 }
 
-func NewMergedSeries(lset labels.Labels, replicas []storage.Series) storage.Series {
+func NewMergedSeries(lset labels.Labels, replicas []storage.Series, f string) storage.Series {
 	return &mergedSeries{
 		lset:     lset,
 		replicas: replicas,
+
+		isCounter: isCounter(f),
 	}
 }
 
 func (m *mergedSeries) Labels() labels.Labels {
 	return m.lset
 }
 func (m *mergedSeries) Iterator(_ chunkenc.Iterator) chunkenc.Iterator {
-	iters := make([]chunkenc.Iterator, 0, len(m.replicas))
+	iters := make([]adjustableSeriesIterator, 0, len(m.replicas))
 	oks := make([]bool, 0, len(m.replicas))
 	for _, r := range m.replicas {
-		it := r.Iterator(nil)
+		var it adjustableSeriesIterator
+		if m.isCounter {
+			it = &counterErrAdjustSeriesIterator{Iterator: r.Iterator(nil)}
+		} else {
+			it = &noopAdjustableSeriesIterator{Iterator: r.Iterator(nil)}
+		}
```

Comment:
You are applying counter adjustments to the raw data samples before deduplication. This is more complicated than applying the adjustments to the single time series after deduplication. The adjustments will intervene with quorum-based deduplication logic. I'm concerned it may introduce other edge cases.

Reply:
Can you add another test case such that one raw time series has a large gap with a reset, but the other two have complete data?
```
replica 0: [[1000, 10], [10000, 8], [11000, 10]]

replica 1: [[1000, 10], [2000, 0], [3000, 1], [4000, 2], [5000, 3], [6000, 4], [7000, 5], [8000, 6], [9000, 7],  [10000, 8], [11000, 10]
replica 2: [[1000, 10], [2000, 0], [3000, 1], [4000, 2], [5000, 3], [6000, 4], [7000, 5], [8000, 6], [9000, 7],  [10000, 8], [11000, 10]
```

Reply:
added the test you suggested above